File: project/cpp/NFA/State/State.cpp
#include "State.h"

State::State(bool accepting, unordered_map<char, State> transitions) {
    this.accepting = accepting;
    this.transitions = transitions;
}

bool State::isAccepting() const {
    return this.accepting;
}

void State::setTransition(char symbol, State destination) {
    this.transitions.at(symbol) = destination;    
}

void State::addTransition(char symbol, State destination) {
    this.transitions[symbol] = destination;
}

File: project/python/lexer/state.py
from collections import defaultdict
from abc import ABC, abstractmethod

class NFAState:

    def __init__(self, accepting=False):
        self.transitions = defaultdict(set)
        self.accepting = accepting

    def add_transition(self, transition_char, target_state):
        self.transitions[transition_char].add(target_state)

    def show(self):
        lines = [f'{id(self)} (A: {self.accepting}):']
        for char, targets in self.transitions.items():
            targets_string = ','.join([str(id(target)) for target in targets])
            lines.append(f'\t{char} -> {targets_string}')

        return '\n'.join(lines)

class DFAState:
    def __init__(self, accepting=False):
        self.transitions = dict()
        self.accepting = accepting

    def set_transition(self, transition_char, target_state):
        self.transitions[transition_char] = target_state

    def show(self):
        lines = [f'{id(self)} (A: {self.accepting}):']
        for char, target in self.transitions.items():
            lines.append(f'\t{char} -> {id(target)}')

        return '\n'.join(lines)
